### ISSUE TITLE
Auto-detect precompute

### ISSUE THREAD

**Issue by tbekolay** [nengo/nengo-loihi#53]
Right now we have users pass in `precompute=True` or `precompute=False` to change how the model is organized under the hood. There's not much guidance as to why they would choose one vs the other, so it feels odd having this be a user choice.

Ideally, whether the model is to be precomputed or not should be automatically determined, to the point that it is not user configurable.

The next level of this would be to allow the user to specify a preference, but consider that a suggestion and switch when we have reason to believe that a model can be precomputed (or not).

The smallest change to the current situation that would be helpful would be to detect the situation that results in the `Cannot precompute input, as it is dependent on output` exception and automatically switch to `precompute=False` in that situation.

**Comment by tbekolay** [comment]
I think nodes that can be precomputed are the rare cases, and in general would be pretty easy to whitelist and otherwise assume it can't be precomputed. So things like the processes in `nengo/processes.py` and a few other known ones we can whitelist, otherwise assume it can't be precomputed.

The annoying (and obviously very common) case are functions. I think that's the case where manually specifying through the config system makes sense, and assuming that they're not precomputable otherwise. But yeah, I just definitely see manually specifying `precompute` as a tricky thing to teach users, so we should try to help as much as possible.

**Comment by tcstewar** [comment]
> I think nodes that can be precomputed are the rare cases, and in general would be pretty easy to whitelist and otherwise assume it can't be precomputed. So things like the processes in nengo/processes.py and a few other known ones we can whitelist, otherwise assume it can't be precomputed.

I completely agree that that's the assumption we should make (any Node that's an unknown function and is not on a whitelist should be assumed to not be precomputable).   So are you envisioning an eventual system where there's a `precompute=None`, which is the default and the builder chooses, and it only chooses `precompute=True` if all of the Nodes are on a whitelist?   If so, then I think the "smallest change to the current situation that would be helpful" would just be switching the default from `precompute=True` to `precompute=False`.   Or is there a reason that we don't want to change that default back to what it was?

**Comment by tbekolay** [comment]
I'm envisioning what I said in the original post, that

1. ideally there's no `precompute` arg and we just figure it out; if that's not possible
2. allow the `precompute` arg as a suggestion that we can override (i.e., if they say don't precompute but we determine that we definitely without a shadow of a doubt can, we go ahead and precompute it); and if that's not possible then
3. do what we do now, but instead of raising an exception when we determine `precompute=True` won't work, we warn and switch it to `precompute=False`.

The main new thing from this thread for me is that instead of using a `precompute` argument to `Simulator` we could instead use the config system to mark nodes as being precomputable. So that would be a variant of 2 and 3 in the list above.

**Comment by tbekolay** [comment]
Dear Github: "does not close #53" should not close #53

**Comment by arvoelke** [comment]
Just adding a discussion point here so that it's not lost: it may be interesting/useful to explore a hybrid combation of `True` and `False`, where the maximal subraph(s) that can be precomputed are automatically identified, and rest is provided online. This might help for real-time applications where some portion of the input is expensive to compute (relative to IO) but can be pre-computed separately using Nengo objects. Considerations might include: how common are such use-cases, when do such use-cases benefit from precomputing some portion of the model, and the trade-offs associated with the overhead / per-byte costs of IO channels. On the last points, it would be good to have systematic profiling benchmarks in place to quantify the cost of real-time IO. This might also depend on multi-chip support, with SNIPS distributed across chips.
